Renaming recurrent layers

rahulbhalley opened the pull request:
This PR suggests the renaming of recurrent neural networks based on the discussion on this [Google groups thread](https://groups.google.com/a/tensorflow.org/forum/#!topic/swift/KjOLKmPVtA8).

1. `SimpleRNNCell` have been renamed to `RNNCell`. This is the vanilla recurrent neural network.
2. `RNNCell` have been renamed to `RNNCellProtocol`. This is the protocol adopted by all recurrent neural networks namely, `RNNCell` (now is a structure), `LSTMCell`, and `GRUCell`.

dan-zheng commented on Sources/TensorFlow/Layers/Recurrent.swift:
```diff
@@ -434,5 +434,5 @@ public struct RNN<Cell: RNNCell>: Layer {
 extension RNN: Equatable where Cell: Equatable {}
 extension RNN: AdditiveArithmetic where Cell: AdditiveArithmetic {}
 
-public typealias SimpleRNN<Scalar: TensorFlowFloatingPoint> = RNN<SimpleRNNCell<Scalar>>
```
Note: the correspondence between `SimpleRNN` and `SimpleRNNCell` seems more obvious than `SimpleRNN` and `RNNCell`. The name `SimpleRNNCell` also has precedent in [Keras](https://www.tensorflow.org/api_docs/python/tf/keras/layers/SimpleRNNCell).

rahulbhalley replied:
Hmm 🧐, you are right.

rahulbhalley commented:
@saeta thanks for positive response! The idea of dividing this into two PRs look good to me.

> Rename `RNNCell` to `RNNCellProtocol`. Add a deprecated `typealias RNNCell = RNNCellProtocol`. This ensures backwards compatible continuity.

Following your suggestion, I'll add a backward compatibility `typealias RNNCell = RNNCellProtocol` in this PR.

> Note: the correspondence between `SimpleRNN` and `SimpleRNNCell` seems more obvious than `SimpleRNN` and `RNNCell`. The name `SimpleRNNCell` also has precedent in [Keras](https://www.tensorflow.org/api_docs/python/tf/keras/layers/SimpleRNNCell).

Following @dan-zheng suggestion from above quote I've renamed `RNNCell` to `SimleRNNCell` for now (but we can change if design discussion converges to use `RNNCell` decision and other corresponding changes).

For now I'll open a draft PR to remove `typealias` introduced for backward compatibility here.

rahulbhalley commented:
@saeta thanks for positive response! The idea of dividing this into two PRs look good to me.

> Rename `RNNCell` to `RNNCellProtocol`. Add a deprecated `typealias RNNCell = RNNCellProtocol`. This ensures backwards compatible continuity.

Following your suggestion, I've added a backward compatibility `typealias RNNCell = RNNCellProtocol` in this PR.

> Note: the correspondence between `SimpleRNN` and `SimpleRNNCell` seems more obvious than `SimpleRNN` and `RNNCell`. The name `SimpleRNNCell` also has precedent in [Keras](https://www.tensorflow.org/api_docs/python/tf/keras/layers/SimpleRNNCell).

Following @dan-zheng suggestion from above quote I've renamed `RNNCell` to `SimleRNNCell` for now (but we can change if design discussion converges to use `RNNCell` decision and other corresponding changes).

ematejska commented:
This is scheduled to be discussed in this Fridays Design (2/28) meeting.

Shashi456 commented:
Having taken a look at other frameworks in deep learning this is what I have found:
- The base class in Keras analogous to `RNNCellProtocol` is `AbstractRNNCell`, take a look [here](https://www.tensorflow.org/api_docs/python/tf/keras/layers/AbstractRNNCell). In accordance with that, actually `AbstractRNNCell` does fit in for our use case as well, while PyTorch chooses `RNNBase` as the base class instead.  
- Chiming in yesterday's ideas thrown around in the meeting, there was `neuralRecurrenceProtocol` and `recurrenceProtocol`.

dan-zheng commented:
> * The base class in Keras analogous to `RNNCellProtocol` is `AbstractRNNCell`, take a look [here](https://www.tensorflow.org/api_docs/python/tf/keras/layers/AbstractRNNCell). In accordance with that, actually `AbstractRNNCell` does fit in for our use case as well, while PyTorch chooses `RNNBase` as the base class instead.

As terms, "abstract" and "base" are both inappropriate in my opinion because they imply "abstract class" and "base class" (superclass), which involve Python class inheritance. In Swift, we're using *protocols*, not *class inheritance*.

`RNNBase` as the name for a "recurrent neural network cell" protocol doesn't quite make sense.

> * Chiming in yesterday's ideas thrown around in the meeting, there was `neuralRecurrenceProtocol` and `recurrenceProtocol`.

Thanks for sharing!

During the meeting, Martin shared that `RNNCellProtocol` makes more sense than `RecurrentCellProtocol`.
* It is not that the case that these cell are inherently recursive at all: they take an input and a state and produce an output and a state.
* It is the neural network wrapping a cell (the `RNN<Cell>` wrapper) that is actually recursive - it recursively feeds the output state of a cell to itself as the next input state.

---

All said, I like `RNNCellProtocol`. I'm not sure there exists a better name than `RNNCellProtocol`, but I'm always happy to entertain ideas!

Shashi456 commented:
+1 on `RNNCellProtocol`, given the arguments.

rahulbhalley commented:
I like the name `RNNCellProtocol` because it brings the clarity to naming: main reason why this PR was opened.

dan-zheng commented:
All Swift code in [tensorflow/swift-apis](https://github.com/tensorflow/swift-apis) was automatically formatted via [`swift-format`](https://github.com/apple/swift-format) in https://github.com/tensorflow/swift-apis/pull/711.

[Formatting configuration here](https://github.com/tensorflow/swift-apis/blob/master/.swift-format).
One notable change is that we're now using 2-space indentation, which is consistent with:
- [Google Swift Style Guide](https://google.github.io/swift)
- The Swift standard library style
- The style of [apple/swift-format](https://github.com/apple/swift-format)

---

As a result, this pull request now has merge conflicts due to code formatting changes. Please resolve the conflicts if you'd like us to re-review your patch! If you haven't already, please consider installing [`swift-format`](https://github.com/apple/swift-format) so you can auto-format code by yourself in the future!

saeta commented:
LGTM; thanks @dan-zheng, @dabrahams, and @rahulbhalley ! (And thank you for including the backwards compatible typealiases too.)

rahulbhalley commented:
These names are beautiful.

I was skeptical about the following earlier but I agree with your point.
> The name RNN implies "neural network", though BasicRNN is actually a layer. This seems okay though, because deep learning users understand that types called "RNN" are actually layers intended for use as building blocks.

And thank you @dabrahams. Sir I'm a fan of your "Protocol-Oriented Programming in Swift" talk at WWDC 15. Your storytelling was fantastic! 🤩 I hope your "friend Crusty" will admire differentiable type system in Swift! 😃

dabrahams commented:
You're very welcome!  I hope you'll also watch Embracing Algorithms
<https://developer.apple.com/videos/play/wwdc2018/223/>, and that you'll
use my talks to kindle an interest in the work of Alexander Stepanov
<http://stepanovpapers.com/#Algorithmic_Journeys>, to whom I owe most of
the inspiration.  Sorry, couldn't resist.

On Fri, Mar 6, 2020 at 1:47 AM ਰਾਹੁਲ ਭੇਲੇ <notifications@github.com> wrote:

> And thank you @dabrahams <https://github.com/dabrahams>. Sir I'm a fan of
> your "Protocol-Oriented Programming in Swift" talk at WWDC 15. Your
> storytelling was fantastic! 🤩 I hope your "friend Crusty" will admire
> differentiable type system in Swift! 😃
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/swift-apis/pull/693?email_source=notifications&email_token=AAAKYIMODWNCQRZL7USS6S3RGDBC3A5CNFSM4KZYOTD2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEOAXZHI#issuecomment-595688605>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAKYIMQNEJXO6B7O472BQ3RGDBC3ANCNFSM4KZYOTDQ>
> .
>


-- 
-Dave

rahulbhalley commented:
What are further plans on this PR? @dan-zheng @dabrahams

dan-zheng commented:
I fixed merge conflicts. LGTM now!
